### PR TITLE
Require env var for custom git fetch

### PR DESCRIPTION
**Pull request by daxmc99**
Move `customGitFetch` configuration from site-admin to a local file.

## Test plan

Manual review 
Manual testing

### CODE REVIEW

**Review by andreeleuterio — APPROVED** [review]

**Comment by evict** [comment]
Thanks @andreeleuterio for proposing to moving it to a file! Makes things a lot easier.

File: cmd/gitserver/server/customfetch_test.go [hunk 1]
```diff
@@ -82,4 +92,8 @@ func TestCustomGitFetch(t *testing.T) {
 			t.Errorf("URL %q: %v", test.url, diff)
 		}
 	}
+
+	t.Cleanup(func() {
+		os.Remove(customGitFetchCmdConf)
+	})
```
Comment (daxmc99) [inline cmd/gitserver/server/customfetch_test.go:98]: `t.Cleanup` should be called near the resources you want to remove. It should be called around line 54

File: cmd/gitserver/server/customfetch.go [hunk 2]
```diff
@@ -2,27 +2,39 @@ package server
...
 
-func buildCustomFetchMappings(c []*schema.CustomGitFetchMapping) map[string][]string {
-	if c == nil {
+func customGitFetch() map[string][]string {
+	if customGitFetchCmdConf == "" {
+		return map[string][]string{}
+	}
+
+	r, err := ioutil.ReadFile(customGitFetchCmdConf)
+
```
Comment (unknwon) [inline cmd/gitserver/server/customfetch.go:24]: nit

```suggestion
```

File: cmd/gitserver/server/customfetch.go [hunk 3]
```diff
@@ -2,27 +2,39 @@ package server
...
+	if err != nil {
+		return map[string][]string{}
+	}
+	var cc []*schema.CustomGitFetchMapping
+
+	err = json.Unmarshal(r, &cc)
+
```
Comment (unknwon) [inline cmd/gitserver/server/customfetch.go:31]: nit
```suggestion

	var cc []*schema.CustomGitFetchMapping
	err = json.Unmarshal(r, &cc)
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -27,6 +27,7 @@ All notable changes to Sourcegraph are documented in this file.
 ### Changed
 
 - **IMPORTANT: Search queries with patterns surrounded by** `/.../` **will now be interpreted as regular expressions.** Existing search links or code monitors are unaffected. In the rare event where older links rely on the literal meaning of `/.../`, the string will be automatically quoted it in a `content` filter, preserving the original meaning. If you happen to use an existing older link and want `/.../` to work as a regular expression, add `patterntype:standard` to the query. New queries and code monitors will interpret `/.../` as regular expressions. [#38141](https://github.com/sourcegraph/sourcegraph/pull/38141).
+- Custom Git fetch configuration `experimentalFeatures.customGitFetch` is now deprecated. The configuration is now moved to a file, use environment variable `CUSTOM_GIT_FETCH_CONF` to set the config file to read the settings from. [#39630](https://github.com/sourcegraph/sourcegraph/pull/39630)
```
Comment (unknwon) [inline CHANGELOG.md:30]: nit: it is actually unused, "deprecated" sounds like people can still live with the old approach (i.e. change through site config)

File: cmd/gitserver/server/customfetch_test.go [hunk 5]
```diff
@@ -42,6 +44,22 @@ func TestCustomGitFetch(t *testing.T) {
...
 	}
 
+	c, err := json.Marshal(mappings)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// write config file to disk
+	err = ioutil.WriteFile(customGitFetchCmdConf, c, 0644)
+
```
Comment (unknwon) [inline cmd/gitserver/server/customfetch_test.go:54]:
```suggestion
```

**Review by unknwon — APPROVED** [review]
Thanks and LGTM other than some world-class formatting issues 😛

File: CHANGELOG.md [hunk 6]
```diff
@@ -27,6 +27,7 @@ All notable changes to Sourcegraph are documented in this file.
 ### Changed
 
 - **IMPORTANT: Search queries with patterns surrounded by** `/.../` **will now be interpreted as regular expressions.** Existing search links or code monitors are unaffected. In the rare event where older links rely on the literal meaning of `/.../`, the string will be automatically quoted it in a `content` filter, preserving the original meaning. If you happen to use an existing older link and want `/.../` to work as a regular expression, add `patterntype:standard` to the query. New queries and code monitors will interpret `/.../` as regular expressions. [#38141](https://github.com/sourcegraph/sourcegraph/pull/38141).
+- Custom Git fetch configuration `experimentalFeatures.customGitFetch` is now unused. The configuration has been moved to a file and no longer uses the site-admin settings. Use environment variable `CUSTOM_GIT_FETCH_CONF` to set the config file to read the settings from. [#39630](https://github.com/sourcegraph/sourcegraph/pull/39630).
```
Comment (DaedalusG) [inline CHANGELOG.md:30]: Assuming the `CUSTOM_GIT_FETCH_CONF` env var is to be set on the `sourcegraph-frontend` deployment? Does this env var need to be set on any other deployments?

**Review by DaedalusG — COMMENTED** [review]
Assuming the CUSTOM_GIT_FETCH_CONF env var is to be set on the sourcegraph-frontend deployment? Does this env var need to be set on any other deployments?

File: CHANGELOG.md (hunk 6 above)
Comment (evict) [reply CHANGELOG.md:30]: The env var needs to be set on the `gitserver`! Doesn't work from the `frontend` deployment.

Comment (DaedalusG) [reply CHANGELOG.md:30]:
```suggestion
- Custom Git fetch configuration `experimentalFeatures.customGitFetch` is now unused. The configuration has been moved to a file and no longer uses the site-admin settings. In gitserver, use environment variable `CUSTOM_GIT_FETCH_CONF` to set the config file to read the settings from. [#39630](https://github.com/sourcegraph/sourcegraph/pull/39630).
```

**Comment by DaedalusG** [comment]
Moved the docs updates from a separate branch to this one 👍, please take one more look at this since I've created a dedicated usage doc

**Review by unknwon — APPROVED** [review]

**Review by DaedalusG — APPROVED** [review]

**Comment by daxmc99** [comment]
@evict @andreeleuterio Is there a reason we moved this to a file? 
I actually think this UX is noticeably harder for the critical customer who uses this. 
I don't think this can be complete until we provide precise instructions for how to use this in a k8s deployment environment. 
I'm happy to do that but I'm not fully understanding what moving from a config option to a file gives us. 

It seems simpler to keep the config in the site-config and leverage an infrastructure-level config option to tune it.

**Review by DaedalusG — CHANGES_REQUESTED** [review]
Holding this till we can provide easy instructions of how to mount the config file

File: cmd/gitserver/server/customfetch_test.go [hunk 7]
```diff
@@ -23,6 +22,9 @@ func TestEmptyCustomGitFetch(t *testing.T) {
 }
 
 func TestCustomGitFetch(t *testing.T) {
+	// mock value for test
+	customGitFetchCmdConf = "/tmp/customGitConfig.json"
```
Comment (keegancsmith) [inline cmd/gitserver/server/customfetch_test.go:26]: /tmp is not always a writeable tmp dir. Luckily go provides something which will return a path for you and clean it up when the test is done:
```suggestion
	customGitFetchCmdConf = filepath.Join(t.TempDir(), "customGitConfig.json")
```

File: schema/site.schema.json [hunk 8]
```diff
@@ -303,11 +303,11 @@
           }
         },
         "customGitFetch": {
-          "description": "JSON array of configuration that maps from Git clone URL domain/path to custom git fetch command.",
+          "description": "JSON array of configuration that maps from Git clone URL domain/path to custom git fetch command. DEPRECATED: CONFIGURATION MOVED TO A LOCAL FILE USE ENVIRONMENT VARIABLE `CUSTOM_GIT_FETCH_CONF` TO SET FILE PATH.",
```
Comment (keegancsmith) [inline schema/site.schema.json:306]: nit: put the deprecated notice at the start of description. Also say the variable is now ignored.

**Review by keegancsmith — COMMENTED** [review]
What is the original motivation for this? TBH what we are doing now is just more clunky and lives outside of things like being able to use the migrator/etc for it.

Generally when deprecating something we give a release or two before making it non functional.

**Comment by andreeleuterio** [comment]
@keegancsmith you can find more context [here](https://github.com/sourcegraph/security-issues/issues/229). tl;dr is that this can't be a site-admin setting in the context of Managed Instances.

File: schema/site.schema.json [hunk 9]
```diff
@@ -303,11 +303,11 @@
           }
         },
         "customGitFetch": {
-          "description": "JSON array of configuration that maps from Git clone URL domain/path to custom git fetch command.",
+          "description": "DEPRECATED: THIS VALUE WILL IS IGNORED. CONFIGURATION MOVED TO A LOCAL FILE USE ENVIRONMENT VARIABLE `CUSTOM_GIT_FETCH_CONF` TO SET FILE PATH. JSON array of configuration that maps from Git clone URL domain/path to custom git fetch command.",
```
[the remark this answers is not in the document]
Comment (keegancsmith) [reply schema/site.schema.json]: Can't we still read this value if the envvar is not set for 2 releases just so we don't break the customers we know use this? Additionally I'd document that the envvar needs to be set on the gitserver container.

**Comment by evict** [comment]
Closing this MR in favour of a simpler solution: https://github.com/sourcegraph/sourcegraph/pull/42704. This makes sure the customer can keep using it while others have the option of disabling it.

File: schema/site.schema.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (daxmc99) [reply schema/site.schema.json]: Please provide feedback on this new PR https://github.com/sourcegraph/sourcegraph/pull/42704

**Pull request closed without merging**